skate-plugin/src/main/kotlin/com/slack/sgp/intellij/featureflags/FeatureFlagAnnotator.kt

```diff
@@ -0,0 +1,101 @@
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.sgp.intellij.featureflags
+
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.ide.BrowserUtil
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.ExternalAnnotator
+import com.intellij.lang.annotation.HighlightSeverity
+import com.intellij.openapi.components.service
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.slack.sgp.intellij.SkatePluginSettings
+import com.slack.sgp.intellij.TEST_KOTLIN_LANGUAGE_ID_KEY
+import com.slack.sgp.intellij.util.isLinkifiedFeatureFlagsEnabled
+import java.net.URI
+import org.jetbrains.kotlin.idea.KotlinLanguage
+
+class FeatureFlagAnnotator : ExternalAnnotator<List<FeatureFlagSymbol>, List<FeatureFlagSymbol>>() {
+
+  override fun collectInformation(file: PsiFile): List<FeatureFlagSymbol> {
+    if (
+      !file.project.isLinkifiedFeatureFlagsEnabled() ||
+        !isKotlinFile(file) ||
+        !isKotlinFeatureFile(file)
+    ) {
+      return emptyList()
+    }
+    val flags = FeatureFlagExtractor.extractFeatureFlags(file)
+    return transformToFeatureFlagSymbols(file, flags)
+  }
+
+  override fun doAnnotate(collectedInfo: List<FeatureFlagSymbol>): List<FeatureFlagSymbol> =
+    collectedInfo
+
+  override fun apply(
+    file: PsiFile,
+    annotationResult: List<FeatureFlagSymbol>,
+    holder: AnnotationHolder
+  ) {
+    for (symbol in annotationResult) {
+      val message = "Open at: ${symbol.url}"
+      holder
+        .newAnnotation(HighlightSeverity.INFORMATION, "Open for more details.")
+        .range(symbol.element)
+        .needsUpdateOnTyping(true)
+        .withFix(UrlIntentionAction(message, symbol.url))
+        .create()
+    }
+  }
+
+  private fun transformToFeatureFlagSymbols(
+    psiFile: PsiFile,
+    flags: List<PsiElement>?
+  ): List<FeatureFlagSymbol> {
+    val baseUrl = psiFile.project.service<SkatePluginSettings>().featureFlagBaseUrl.orEmpty()
+    return flags.orEmpty().map { flag -> FeatureFlagSymbol(flag, "$baseUrl?q=${flag.text}") }
+  }
+
+  private fun isKotlinFile(psiFile: PsiFile): Boolean =
+    psiFile.language.id == KotlinLanguage.INSTANCE.id ||
+      psiFile.getUserData(TEST_KOTLIN_LANGUAGE_ID_KEY) == KotlinLanguage.INSTANCE.id
+
+  private fun isKotlinFeatureFile(psiFile: PsiFile): Boolean = psiFile.name.endsWith("Feature.kt")
+}
+
+class FeatureFlagSymbol(val element: PsiElement, val url: String)
```

Comment:
This should use a stronger type than `PsiElement`

Reply:
[ec81f44](https://github.com/slackhq/slack-gradle-plugin/pull/576/commits/ec81f4417cd0409bed9ab368d3bdb606b12ea717#diff-0073d4ea07ce931d766fbad1989d075c4289e50411408234e22459bde49a29bdR76)